src/composition/models/stacking.jl

```
@@ -0,0 +1,282 @@
+############################################
+################ Structures ################ 
+############################################
+
+function is_glb(potential_glb, models)
+    for model in models
+        if !(potential_glb <: input_scitype(model))
+            return false
+        end
+    end
+    return true
+end
+
+function glb(models)
+    for model in models
+        potential_glb = input_scitype(model)
+        if is_glb(potential_glb, models)
+            return potential_glb
+        end
+    end
+    return Unknown
+end
+
+
+mutable struct DeterministicStack{modelnames, input_scitype, target_scitype} <: DeterministicComposite
+   models::Vector{Supervised}
+   metalearner::Deterministic
+   resampling
+   function DeterministicStack(modelnames, models, metalearner, resampling)
+        target_scitype = MMI.target_scitype(metalearner)
```

I think I was a bit sloppy in the design brief. It seems to me we should probably make the stack `target_scitype` the greatest lower bound of the target scitypes of the base learners and metalearner, no? At present there is nothing stopping the user specifying a regressor as a base model when the metalearner is a classifier, which doesn't make sense. 

To this end, you may want to refactor `glb(models)` to new signature `glb(types...)`.

---

I think it makes sense to define the `target_scitype` as a glb of both models in the library and the metalearner. The question is, should we then raise if the `target_scitype` is then Unknown?  

This makes me realize that changing a model or the metalearner by `setproperty!` should then change the `input_scitype` or `target_scitype` of the stack. Eventually I guess it should change the parametric type of the stack too? Which I think is not what we want to do (not even sure if this is possible?). Maybe we should prevent from using `setproperty!` on the models and the metalearner? It should still be possible to change hyperparameters of the models themselves.

---

No a type parameter cannot be changed, which I don't think is too big a restriction here. We should however allow mutations of component models that are consistent with the original stack scitypes (I suppose  "more permissive" model mutations could be allowed, but this is probably too sneaky). `clean!` should check consistency and throw an error otherwise. I'd throw an error in that case. Note that 99% of the time computing scitypes is not expensive, so recomputing every `fit!` call is no big deal, in my opinion.

If the target scitype or input scitype *computed at construction* is `Unknown`, then I'd issue `@warn`.

---

Ok, I think I've addressed the issue